core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1603,6 +1632,8 @@ class Log(@volatile var dir: File,
    */
   private[log] def replaceSegments(newSegment: LogSegment, oldSegments: Seq[LogSegment], isRecoveredSwapFile: Boolean = false) {
     lock synchronized {
+      if (dir == null)
+        throw new KafkaStorageException(s"The log for partition $topicPartition is offline")
```

Comment:
Hmm, since LogCleaner.doWork() doesn't catch KafkaStorageException, this will kill the log cleaner thread, which is not what we want.

Reply:
Good point. I missed this. I have updated the patch to catch and swallow KafkaStorageException in `LogCleaner.cleanOrSleep()`.